### ISSUE TITLE
Unable to generate public/private keypair for YubiKey U2F via OpenSSH: FIDO_ERR_RX

### ISSUE THREAD

**Issue by bturley2** [Yubico/libfido2#188]
This issue is very similar to #145 

I'm running MacOS with the latest version of OpenSSH updated to:
`OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020`

My current version of libfido2 is: `1.4.0` installed via Homebrew.

My Yubikey model is: Yubikey C Nano FIPS
My Yubikey firmware is: 4.4.5

When I attempt to run:
`ssh-keygen -t ecdsa-sk -vv`

I'm met with the same error of:

```
Generating public/private ecdsa-sk key pair.
You may need to touch your authenticator to authorize key generation.
debug1: start_helper: starting /usr/local/libexec/ssh-sk-helper
debug1: sshsk_enroll: provider "internal", device "(null)", application "ssh:", userid "(null)", flags 0x01, challenge len 0
debug1: sshsk_enroll: using random challenge
debug1: ssh_sk_enroll: using device IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS07@14200000/Yubikey 4 OTP+U2F+CCID@14200000/IOUSBHostInterface@1/AppleUserUSBHostHIDDevice
debug1: ssh_sk_enroll: fido_dev_make_cred: FIDO_ERR_RX
debug1: sshsk_enroll: provider "internal" returned failure -1
debug1: ssh-sk-helper: Enrollment failed: invalid format
debug1: ssh-sk-helper: reply len 8
debug1: client_converse: helper returned error -4
Key enrollment failed: invalid format
```

Based on #145 a fix for this was supposed to be included in libfido2 1.4.0, but as I'm currently running that version of libfido2 it appears as though the problem has not been resolved.

Is anyone able to provide help for this issue?

**Comment by martelletto** [comment]
A fix for this issue was _not_ included in 1.4.0. I have updated my comment in #145. Thanks for making me realise my mistake. :) Homebrew did, however, pull the fix: https://github.com/Homebrew/homebrew-core/blob/master/Formula/libfido2.rb#L21-L25.

Can you `brew upgrade libfido2` and ascertain you are running with libfido2 1.4.0_2?

-p.

**Comment by bturley2** [comment]
Running:
`brew upgrade libfido2`

provides the following results:
```
$ brew upgrade libfido2
Updating Homebrew...
Warning: libfido2 1.4.0_2 already installed
```

**Comment by martelletto** [comment]
Thanks; let's dig into it then. As a point of reference, `ssh-keygen -t ecdsa-sk -vv` works for me on a Yubikey 4 FIPS with firmware 4.4.2 (also on macOS). Are you building ssh from source? If so, can you [enable SK_DEBUG in sk-usbhid.c](https://github.com/openssh/openssh-portable/blob/master/sk-usbhid.c#L58)? Otherwise, can you build libfido2 from source and try to run examples/cred with the environment variable FIDO_DEBUG set to 1? This should work:
```
$ git clone https://github.com/yubico/libfido2
$ cd libfido2
$ export PKG_CONFIG_PATH=$(brew --prefix openssl)/lib/pkgconfig
$ mkdir build
$ cd build
$ cmake ..
$ examples/manifest
$ export FIDO_DEBUG=1
$ examples/cred <path_to_authenticator>
```

**Comment by bturley2** [comment]
(Please have some patience with me, I'm fairly new to building from source.)

I had tried building ssh from source earlier, but I just went ahead and reinstalled/overwrote the previous installation of OpenSSH using brew.
```
brew install openssh
brew link --overwrite openssh
```

What do you mean on that last command by `<path_to_authenticator>`?

**Comment by martelletto** [comment]
By `<path_to_authenticator>` I mean the path produced by `fido2-token -L`:
```
$ fido2-token -L
IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS04@14400000/YubiKey FIDO@14400000/IOUSBHostInterface@0/AppleUserUSBHostHIDDevice: vendor=0x1050, product=0x0402 (Yubico YubiKey FIDO)
```
In this case, the path to the authenticator is the string `IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS04@14400000/YubiKey FIDO@14400000/IOUSBHostInterface@0/AppleUserUSBHostHIDDevice`.

**Comment by bturley2** [comment]
I may be formatting the last command incorrectly. I'm receiving the following results:

```
$ examples/cred IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS07@14200000/Yubikey 4 OTP+U2F+CCID@14200000/IOUSBHostInterface@1/AppleUserUSBHostHIDDevice
usage: cred [-t ecdsa|rsa|eddsa] [-k pubkey] [-ei cred_id] [-P pin] [-T seconds] [-hruv] <device>
```

It seems as though the command expects additional inputs, so I tried running the same command with the `-t ecdsa` tag and got the same results:
```
$ examples/cred -t ecdsa IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS07@14200000/Yubikey 4 OTP+U2F+CCID@14200000/IOUSBHostInterface@1/AppleUserUSBHostHIDDevice
usage: cred [-t ecdsa|rsa|eddsa] [-k pubkey] [-ei cred_id] [-P pin] [-T seconds] [-hruv] <device>
```

**Comment by martelletto** [comment]
You need to quote the path. Otherwise the shell will break it into multiple arguments, since it contains spaces.
``` 
$ examples/cred -t ecdsa "IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS07@14200000/Yubikey 4 OTP+U2F+CCID@14200000/IOUSBHostInterface@1/AppleUserUSBHostHIDDevice"
```

**Comment by bturley2** [comment]
Thanks for the clarification.

Running that command provides the following output:
```
$ examples/cred "IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/XHC1@14/XHC1@14000000/HS07@14200000/Yubikey 4 OTP+U2F+CCID@14200000/IOUSBHostInterface@1/AppleUserUSBHostHIDDevice"
fido_tx: d=0x7fee8070f580, cmd=0x06, buf=0x7fee8070f580, count=8
0000: 35 fc 5a 1a ff 85 7b 8e
fido_rx: d=0x7fee8070f580, cmd=0x06, buf=0x7fee8070f588, count=17, ms=-1
rx_preamble: initiation frame at 0x7ffee850d3b8
0000: ff ff ff ff 86 00 11 35 fc 5a 1a ff 85 7b 8e 00
0016: 0a 00 01 02 04 04 05 05 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx: payload_len=17
fido_rx: buf=0x7fee8070f588, len=17
0000: 35 fc 5a 1a ff 85 7b 8e 00 0a 00 01 02 04 04 05
0016: 05
fido_dev_get_cbor_info_tx: dev=0x7fee8070f580
fido_tx: d=0x7fee8070f580, cmd=0x10, buf=0x7ffee850d417, count=1
0000: 04
fido_dev_get_cbor_info_rx: dev=0x7fee8070f580, ci=0x7fee8050a560, ms=-1
fido_rx: d=0x7fee8070f580, cmd=0x10, buf=0x7ffee850cc10, count=2048, ms=-1
rx_preamble: initiation frame at 0x7ffee850cb28
0000: 00 0a 00 01 bf 00 01 01 00 00 00 00 00 00 00 00
0016: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx_preamble: cid (0x1000a00, 0x1000a00), cmd (0xbf, 0x10)
rx: rx_preamble
fido_dev_get_cbor_info_rx: fido_rx
fido_dev_open_rx: falling back to u2f
fido_tx: d=0x7fee8070f580, cmd=0x03, buf=0x7fee806046c2, count=73
0000: 00 01 00 00 00 00 40 f9 64 57 e7 2d 97 f6 bb dd
0016: d7 fb 06 37 62 ea 26 20 44 8e 69 7c 03 f2 31 2f
0032: 99 dc af 3e 8a 91 6b 49 96 0d e5 88 0e 8c 68 74
0048: 34 17 0f 64 76 60 5b 8f e4 ae b9 a2 86 32 c7 99
0064: 5c f3 ba 83 1d 97 63 00 00
fido_rx: d=0x7fee8070f580, cmd=0x03, buf=0x7ffee850cd00, count=2048, ms=-1
rx_preamble: initiation frame at 0x7ffee850cbe8
0000: 00 0a 00 01 83 00 02 63 c0 00 00 00 00 00 00 00
0016: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx: payload_len=2
fido_rx: buf=0x7ffee850cd00, len=2
0000: 63 c0
parse_register_reply: unexpected sw
u2f_register: parse_register_reply
cred: fido_makecred: FIDO_ERR_RX (0xfffffffe)
```

**Comment by martelletto** [comment]
Thank you. Do you have a FIPS PIN set?

**Comment by bturley2** [comment]
Yes I do.

**Comment by martelletto** [comment]
That's it, then. libfido2 does not support U2F with PIN. To the best of my knowledge, that functionality sits outside the CTAP2/CTAP1 specs.

**Comment by bturley2** [comment]
So having a PIN associated with my Yubikey does not allow U2F support with OpenSSH?

I have been using U2F for web-login on my Yubikey without issue.

**Comment by martelletto** [comment]
Are you able to use U2F on a browser without a PIN? If so, it should work with OpenSSH/libfido2 as well.

**Comment by bturley2** [comment]
I can't reset my PIN on the original Yubikey that I am experiencing these issues with. However, I also have a Yubikey 5C on hand with which the command works without an issue. When I try setting a PIN on the 5C, it will prompt me for my PIN when I run the `ssh-keygen -t ecdsa-sk` command and work properly:
```
$ ssh-keygen -t ecdsa-sk
Generating public/private ecdsa-sk key pair.
You may need to touch your authenticator to authorize key generation.
Enter PIN for authenticator:
Enter file in which to save the key (/Users/bturley-mac/.ssh/id_ecdsa_sk):
/Users/bturley-mac/.ssh/id_ecdsa_sk already exists.
Overwrite (y/n)? y
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in /Users/bturley-mac/.ssh/id_ecdsa_sk
Your public key has been saved in /Users/bturley-mac/.ssh/id_ecdsa_sk.pub
The key fingerprint is:
SHA256:95R6ZRlm8LUXPDK3lbo5LrnYWvxog+ZUawD6QYWr84U bturley-mac@bturley-mac-mac
The key's randomart image is:
+-[ECDSA-SK 256]--+
|        ..  . ..o|
|       ..    = *+|
|       o.     X.=|
|      o..    = +.|
|     ...S...o *  |
|     o.E.o++.*   |
|      o...o== .  |
|       ..o=*o.   |
|        o+o++.   |
+----[SHA256]-----+
```

**Comment by bturley2** [comment]
The Yubikey C Nano FIPS that I was originally having issues with was sent to me pre-configured by my work. It seems as though there is an admin password set on the device, which appears to be different from the PIN you can set for OTP/PIV functionality etc. (Based on the information from[ this page.](https://developers.yubico.com/PIV/Introduction/Admin_access.html))

Since I need the device configured as is for work, I'm going to go ahead and close this issue. 

For anyone who encounters this issue in the future, Yubico support provided the following comments:
"You mentioned it was a FIPS device - have you locked the U2F Sub-Module with an admin key? If so, the YubiKey will need to be unlocked using ykman prior to registering a new SSH key."

Thank you @martelletto for all your assistance.

**Comment by jabberwock** [comment]
TL;DR for me to get this working:

```
brew install openssh
brew install libfido2
brew install install libsk-libfido2
```
